core/src/types/storage.rs

```diff
@@ -1151,6 +1154,147 @@ pub struct PrefixValue {
     pub value: Vec<u8>,
 }
 
+/// Container of all Ethereum event queues.
+#[derive(Default, Debug, BorshSerialize, BorshDeserialize)]
+pub struct EthEventsQueue {
+    /// Queue of transfer to Namada events.
+    pub transfers_to_namada: InnerEthEventsQueue<TransfersToNamada>,
+    // TODO: add queue of update whitelist events
+}
+
+/// A queue of confirmed Ethereum events to be processed in order.
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub struct InnerEthEventsQueue<E> {
+    next_nonce_to_process: Uint,
+    inner: VecDeque<E>,
+}
+
+impl<E: GetEventNonce> InnerEthEventsQueue<E> {
+    /// Return an Ethereum events queue starting at the specified nonce.
+    pub fn new_at(next_nonce_to_process: Uint) -> Self {
+        Self {
+            next_nonce_to_process,
+            ..Default::default()
+        }
+    }
+}
+
+impl<E: GetEventNonce> Default for InnerEthEventsQueue<E> {
+    fn default() -> Self {
+        Self {
+            next_nonce_to_process: 0u64.into(),
+            inner: Default::default(),
+        }
+    }
+}
+
+/// Draining iterator over a queue of Ethereum events.
+pub struct EthEventsQueueIter<'queue, E> {
+    current_nonce: Uint,
+    queue: &'queue mut InnerEthEventsQueue<E>,
+}
+
+impl<E> Drop for EthEventsQueueIter<'_, E> {
```

Comment:
Are all events in the queue ensured to be valid? Otherwise this may incorrectly commit to the wrong nonce on drop. If that is ensured, can you add a doc comment explaining this invariant must be maintained?

Reply:
it wasn't the case before my last commit, but now I'm pretty certain we maintain this invariant. I'll add the docstring